### PR TITLE
Add flag for disallowing shape or storage change for tensor from .data or .detach()

### PR DESCRIPTION
**Pull request by yf225**
After autograd metadata are moved from `Variable::Impl` to TensorImpl (and `Variable::Impl` no longer wraps an `at::Tensor`), we have to shallow-copy the TensorImpl whenever user wants to create a `Variable` that shares the same storage data with another `Variable`.

However, this has implications for Python call sites that do
```python
tensor.data.in_place_operation_()
```
or
```python
tensor_detached = tensor.detach()
tensor_detached.in_place_operation_()
```
If `in_place_operation_()` only updates the data inside the tensor (such as `zeros_()`), such operation will still work properly; if the in-place operation changes the size, stride or the storage pointer inside the TensorImpl (e.g. `resize_` / `resize_as_` / `set_` / `share_memory_`), such operation on `tensor.data` or `tensor_detached` will no longer update the `tensor`. We will address this inconsistency in the following ways:

1. Add a flag to `Variable` to disallow size/stride/storage_ptr changes from in-place operations such as `resize_` / `resize_as_` / `set_` / `share_memory_`, and set this flag to true when people call `tensor.data` or `tensor.detach()` in Python.
2. Write text in the docs to actively discourage changing the shape or storage of `tensor_detached` and expecting `tensor` to also be updated.

This is the first PR mentioned in https://github.com/pytorch/pytorch/issues/13638.

### CODE REVIEW

File: tools/autograd/gen_variable_type.py [hunk 1]
```diff
@@ -82,6 +82,12 @@
     '_coalesced_',
 }
 
+# Some Variables are not allowed to have their size or storage changed, if they
+# share the same storage with another Variable (e.g. via `.data` or `detach()`)
+CHECK_ALLOW_SIZE_OR_STORAGE_CHANGE = {
+    'set_',
```
Comment (gchanan) [inline tools/autograd/gen_variable_type.py]: I don't think we should hard code these -- can we just make sure that all functions that modify the shape of a TensorImpl go through a single function and check there?
